### DISCUSSION TITLE
Global CODEOWNERs?

### DISCUSSION THREAD

**Discussion by sfc-gh-tochao** [community/community#70685] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Question

### Body

Imagine I have a directory structure that looks like this
```
/foo
/bar
```

and a corresponding CODEOWNERs that looks like
```
/foo @mycompany/foo-owners
/bar @mycompany/bar-owners
```

How can I add a global owners that can approve any PR?  For example if we want to do an emergency rollback.

I tried doing
```
* @mycompany/global-owners
/foo @mycompany/foo-owners
/bar @mycompany/bar-owners
```
but that does not appear to let the global-owners approve PRs because the foo-owners or bar-owners take precedence.

I'm guessing something like
```
/foo @mycompany/foo-owners @mycompany/global-owners
/bar @mycompany/bar-owners @mycompany/global-owners
```
would work but that means the global owners would get pinged on every PR.

What I'm looking for is a mechanism for people in the global owners group to blanket approve PRs, in addition to the existing CODEOWNERs rules.

**Comment by oussamaef06** [comment]
in your scenario, you want to have a mechanism for a "global owners" group to blanket approve PRs in addition to existing CODEOWNERS rules. To achieve this, you can use a combination of CODEOWNERS and branch protection rules in GitHub. Here's how you can set it up:

Define Global Owners in CODEOWNERS:

In your CODEOWNERS file, you can specify a global owner for each directory by using a wildcard pattern *. For example:

* @mycompany/global-owners
/foo @mycompany/foo-owners
/bar @mycompany/bar-owners
This means that the @mycompany/global-owners team will be added as reviewers for every pull request in the repository, but they won't be pinged on every PR by default. They will have the ability to review and approve PRs, but they won't be required to approve every PR.

Branch Protection Rules:

You can set up branch protection rules to require reviews before PRs can be merged. Here's how to configure it:

Go to your GitHub repository.
Click on "Settings."
In the left sidebar, click on "Branches."
Under "Branch protection rules," select the branch you want to protect (e.g., main).
Check the option "Require pull request reviews before merging."
Choose the @mycompany/global-owners team as the required reviewer.
You can also specify the number of required approving reviews if needed.
By setting up branch protection rules as described above, you make it mandatory for PRs to be approved by the @mycompany/global-owners team. This ensures that they have the ability to blanket approve PRs, as you desire. However, they will only be required to approve PRs when the branch protection rule is set to "require pull request reviews before merging."

This approach allows you to maintain the existing CODEOWNERS rules for specific directories while adding a global owner team for broader approvals when necessary, such as emergency rollbacks. The global owners will have the authority to approve PRs without being pinged on every PR, but their approval will be required when branch protection rules dictate.

**Reply by sfc-gh-tochao** [reply]
Is there something special about the wildcard pattern "*"?

I previously had

/foo @mycompany/foo-owners
/foo/bar @mycompany/bar-owners

and only @mycompany/bar-owners could approve changes to /foo/bar.  The foo-owners couldn't even though I expected them to inherit the approvals.  That's why I thought having a generic "* @mycompany/global-owners" wouldn't work.  But if "*" is a special case, then I'll definitely try that.

Thank you!

**Comment by rezafatoolahi** [comment]
Sent from my Galaxy
-------- Original message --------From: oussamaef06 ***@***.***> Date: 30/10/2023  01:21  (GMT+03:30) To: community/community ***@***.***> Cc: Subscribed ***@***.***> Subject: Re: [community/community] Global CODEOWNERs? (Discussion #70685) 
in your scenario, you want to have a mechanism for a "global owners" group to blanket approve PRs in addition to existing CODEOWNERS rules. To achieve this, you can use a combination of CODEOWNERS and branch protection rules in GitHub. Here's how you can set it up:
Define Global Owners in CODEOWNERS:
In your CODEOWNERS file, you can specify a global owner for each directory by using a wildcard pattern *. For example:

@mycompany/global-owners
/foo @mycompany/foo-owners
/bar @mycompany/bar-owners
This means that the @mycompany/global-owners team will be added as reviewers for every pull request in the repository, but they won't be pinged on every PR by default. They will have the ability to review and approve PRs, but they won't be required to approve every PR.

Branch Protection Rules:
You can set up branch protection rules to require reviews before PRs can be merged. Here's how to configure it:
Go to your GitHub repository.
Click on "Settings."
In the left sidebar, click on "Branches."
Under "Branch protection rules," select the branch you want to protect (e.g., main).
Check the option "Require pull request reviews before merging."
Choose the @mycompany/global-owners team as the required reviewer.
You can also specify the number of required approving reviews if needed.
By setting up branch protection rules as described above, you make it mandatory for PRs to be approved by the @mycompany/global-owners team. This ensures that they have the ability to blanket approve PRs, as you desire. However, they will only be required to approve PRs when the branch protection rule is set to "require pull request reviews before merging."
This approach allows you to maintain the existing CODEOWNERS rules for specific directories while adding a global owner team for broader approvals when necessary, such as emergency rollbacks. The global owners will have the authority to approve PRs without being pinged on every PR, but their approval will be required when branch protection rules dictate.
